File: Python/stage_IV.py
# Andoni Garcia. 2014.
# My source code for Stage IV: The Dating Game

from urllib.request import Request, urlopen
from json import dumps, loads
import time
import datetime

token = "yicZ5DfAT6"
response = None
newTime = None

def GetTime():
	global token, response
	url = "http://challenge.code2040.org/api/time"
	values = {'token':token}
	JSONData = dumps(values).encode('utf-8')
	req = Request(url, data = JSONData, headers = {'content-type':'application/json'})
	resp = urlopen(req)
	response = loads(resp.read().decode('utf-8'))['result']
	print("Your interval is: " + str(response['interval']))
	print("Your datestamp is: " + response['datestamp'])

# I could not think of any other way to go about it
def AddTime():
	global response, newTime
	toDateTime = datetime.datetime(*(time.strptime(response['datestamp'], "%Y-%m-%dT%H:%M:%S.%fZ")[0:6]))
	newTime = toDateTime + datetime.timedelta(0, response['interval'])
	print("Your new time is: " + newTime.isoformat(sep="T")+'.000Z')

def PostTime():
	global token, newTime
	url = "http://challenge.code2040.org/api/validatetime"
	values = {'token':token,
			  'datestamp':(newTime.isoformat(sep="T")+'.000Z')}
	JSONData = dumps(values).encode('utf-8')
	req = Request(url, data = JSONData, headers = {'content-type':'application/json'})
	resp = urlopen(req)
	result = loads(resp.read().decode('utf-8'))['result']
	print("Result: " + result)

def StageIV():
	GetTime()
	AddTime()
	PostTime()

StageIV()

File: JavaScript/stage_IV.js
// Andoni Garcia. 2014.
// My source code for Stage IV: The Dating Game

var token = "yicZ5DfAT6"; 
var xmlhttp = new XMLHttpRequest();
var response;
var newTime;


function GetTime(){
	var url = "http://challenge.code2040.org/api/time";
	var values = {"token":token};
	var JSONData = JSON.stringify(values);
	xmlhttp.onreadystatechange = function(){
		if (xmlhttp.readyState == 4 && xmlhttp.status == 200){
			response = JSON.parse(xmlhttp.responseText).result;
			console.log("Your interval is: " + response.interval);
			console.log("Your datestamp is: " + response.datestamp);
		}
	}
	xmlhttp.open("POST", url, false);
	xmlhttp.setRequestHeader("content-type", "application/json");
	xmlhttp.send(JSONData);
}

function AddTime(){
	var interval = response.interval;
	var datestamp = response.datestamp;
	var date = new Date(datestamp);
	var ms = date.getTime() + (interval * 1000);
	newTime = new Date(ms).toISOString();
	console.log("New datestamps is: " + newTime);
}

function PostTime(){
	var url = "http://challenge.code2040.org/api/validatetime";
	var values = {"token":token,
				  "datestamp":newTime};
	var JSONData = JSON.stringify(values);
	xmlhttp.onreadystatechange = function(){
		if (xmlhttp.readyState == 4 && xmlhttp.status == 200){
			var result = JSON.parse(xmlhttp.responseText);
			console.log("Result: " + result.result);
		}
	}
	xmlhttp.open("POST", url, false);
	xmlhttp.setRequestHeader("content-type", "application/json");
	xmlhttp.send(JSONData);
}

// HIGHLY ADVISED TO COMMENT OUT PRINT STATEMENTS BEFORE EXECUTING TIMEFUNCS
function TimeFuncs(){
	GetTime();
	console.time('AddTime');
	AddTime();
	console.timeEnd('AddTime');
}

function StageIV(){
	GetTime();
	AddTime();
	PostTime();
}

StageIV();